api/src/main/java/io/minio/MinioClient.java

```diff
@@ -2589,105 +2700,111 @@ public void composeObject(
+      if (size < PutObjectOptions.MIN_MULTIPART_SIZE
           && sources.size() != 1
           && i != (sources.size() - 1)) {
         throw new IllegalArgumentException(
             "source "
-                + src.bucketName()
+                + src.bucket()
                 + "/"
-                + src.objectName()
+                + src.object()
                 + ": size "
                 + size
                 + " must be greater than "
-                + ObjectWriteArgs.MIN_MULTIPART_SIZE);
+                + PutObjectOptions.MIN_MULTIPART_SIZE);
       }
 
       objectSize += size;
-      if (objectSize > ObjectWriteArgs.MAX_OBJECT_SIZE) {
+      if (objectSize > PutObjectOptions.MAX_OBJECT_SIZE) {
         throw new IllegalArgumentException(
-            "Destination object size must be less than " + ObjectWriteArgs.MAX_OBJECT_SIZE);
+            "Destination object size must be less than " + PutObjectOptions.MAX_OBJECT_SIZE);
       }
 
-      if (size > ObjectWriteArgs.MAX_PART_SIZE) {
-        long count = size / ObjectWriteArgs.MAX_PART_SIZE;
-        long lastPartSize = size - (count * ObjectWriteArgs.MAX_PART_SIZE);
+      if (size > PutObjectOptions.MAX_PART_SIZE) {
+        long count = size / PutObjectOptions.MAX_PART_SIZE;
+        long lastPartSize = size - (count * PutObjectOptions.MAX_PART_SIZE);
         if (lastPartSize > 0) {
           count++;
         } else {
-          lastPartSize = ObjectWriteArgs.MAX_PART_SIZE;
+          lastPartSize = PutObjectOptions.MAX_PART_SIZE;
         }
 
-        if (lastPartSize < ObjectWriteArgs.MIN_MULTIPART_SIZE
+        if (lastPartSize < PutObjectOptions.MIN_MULTIPART_SIZE
             && sources.size() != 1
             && i != (sources.size() - 1)) {
           throw new IllegalArgumentException(
               "source "
-                  + src.bucketName()
+                  + src.bucket()
                   + "/"
-                  + src.objectName()
+                  + src.object()
                   + ": "
                   + "for multipart split upload of "
                   + size
                   + ", last part size is less than "
-                  + ObjectWriteArgs.MIN_MULTIPART_SIZE);
+                  + PutObjectOptions.MIN_MULTIPART_SIZE);
         }
 
         partsCount += (int) count;
       } else {
         partsCount++;
       }
 
-      if (partsCount > ObjectWriteArgs.MAX_MULTIPART_COUNT) {
+      if (partsCount > PutObjectOptions.MAX_MULTIPART_COUNT) {
         throw new IllegalArgumentException(
             "Compose sources create more than allowed multipart count "
-                + ObjectWriteArgs.MAX_MULTIPART_COUNT);
+                + PutObjectOptions.MAX_MULTIPART_COUNT);
       }
     }
 
     if (partsCount == 1) {
       ComposeSource src = sources.get(0);
-      if (headerMap == null) {
-        headerMap = new HashMap<>();
+      Multimap<String, String> headers = HashMultimap.create();
+      if (args.extraHeaders() != null) {
```

Comment:
`extraHeaders` and `extraQueryParams` are always `non-null`. You would need to use them directly.